motionplan/collision.go

```diff
@@ -311,3 +313,99 @@ func createUniqueCollisionMap(geoms []spatial.Geometry) (map[string]spatial.Geom
 	}
 	return geomMap, nil
 }
+
+// CheckPlan checks if obstacles intersect the trajectory of the frame following the plan.
+func CheckPlan(
+	frame referenceframe.Frame,
+	plan [][]referenceframe.Input,
+	obstacles []*referenceframe.GeometriesInFrame,
+	fs referenceframe.FrameSystem,
+) (bool, error) {
+	// ensure that we can actually perform the check
+	if len(frame.DoF()) != len(plan[0]) {
+		return false, errors.New("frame DOF length must match inputs length")
+	}
+	// check if we are working with ptgs
+	ptgProv, ok := frame.(tpspace.PTGProvider)
+	if ok {
+		return checkPtgPlan(frame, plan, obstacles, ptgProv.PTGs())
+	}
+
+	if len(plan) < 2 {
+		return false, errors.New("plan must have at least two elements")
+	}
+
+	// construct planner with collision contraints
+	opt := newBasicPlannerOptions(frame)
+	sf, err := newSolverFrame(fs, frame.Name(), referenceframe.World, nil)
+	if err != nil {
+		return false, err
+	}
+	wrdlst, err := referenceframe.NewWorldState(obstacles, nil)
+	if err != nil {
+		return false, err
+	}
+	collisionConstraints, err := createAllCollisionConstraints(sf, fs, wrdlst, referenceframe.StartPositions(fs), nil)
+	if err != nil {
+		return false, err
+	}
+	for name, constraint := range collisionConstraints {
+		opt.AddStateConstraint(name, constraint)
+	}
+
+	// go through plan and check that we can move from plan[i] to plan[i+1]
+	for i := 0; i < len(plan)-1; i++ {
+		if isValid, fault := opt.CheckSegmentAndStateValidity(
+			&Segment{
+				StartConfiguration: plan[i],
+				EndConfiguration:   plan[i+1],
+				Frame:              frame,
+			},
+			opt.Resolution,
+		); !isValid {
+			return false, fmt.Errorf("found collsion in segment:%v", fault)
+		}
+	}
+	return true, nil
+}
+
+func checkPtgPlan(
+	frame referenceframe.Frame,
+	plan [][]referenceframe.Input,
+	obstacles []*referenceframe.GeometriesInFrame,
+	ptgs []tpspace.PTG,
+) (bool, error) {
```

Comment:
i realize that this function is incomplete.
if we have obstacles which are not in the world frame, we do not correctly perform the collision check.

todo: add this functionality and add corresponding test to tpSpaceRRT_test.go